posts/2018-09-29-monad-transformer-variance.md

```diff
@@ -0,0 +1,161 @@
+```
+
+```tut:book
+val optionNotFound: Option[NotFound] = Some(NotFound(42L))
+
+optionNotFound: Option[Err]
+```
+
+Great. If you want to treat your `Option[Notfound]` as an `Option[Err]`, you are free to.
+
+This is made possible because the `Option` type is declared as `sealed abstract class Option[+A]`, where the `+` in front of the `A` means that it is covariant in the `A` type parameter.
+
+What happens if we try to do the same with `OptionT`?
+
+```tut:silent
+import cats.Eval
+import cats.data.OptionT
+```
+
+```tut:book
+val optionTNotFound: OptionT[Eval, NotFound] = OptionT(Eval.now(optionNotFound))
+```
+
+```tut:book:fail
+optionTNotFound: OptionT[Eval, Err]
+```
+
+The compiler complains that an `OptionT[Eval, NotFound]` is _not_ an `OptionT[Eval, Err]`, but it also suggests that we may be able to fix this by using `+A` like is done with `Option`.
+
+`OptionT` in Cats is defined as:
+
+```scala
+final case class OptionT[F[_], A](value: F[Option[A]])
+```
+
+Let's try to make the suggested change with an experimental `MyOptionT` structure:
+
+```tut:book:fail
+final case class MyOptionT[F[_], +A](value: F[Option[A]])
+```
+
+Now it's complaining that `A` is a covariant type but shows up in an invariant position. We'll discuss more about what this means later in the post, but for now let's just try declaring `F` as covariant:
+
+```tut:silent
+final case class CovariantOptionT[F[+_], +A](value: F[Option[A]])
+```
+
+```tut:book
+val covOptionTNotFound: CovariantOptionT[Eval, NotFound] = CovariantOptionT(Eval.now(optionNotFound))
+
+covOptionTNotFound: CovariantOptionT[Eval, Err]
+```
+
+Woohoo! Problem solved, right?
+
+Well, not exactly. This works great if `F` is in fact covariant, but what if it's not? For example, the JSON library [circe](https://circe.github.io/circe/) has a `Decoder` type that _could_ be covariant but isn't (at least as of circe 0.10.0). With the invariant `OptionT` in Cats we can do something like this:
+
+```tut:silent
+import io.circe.Decoder
+
+def defaultValueDecoder[A](defaultValue: A, optionDecoder: Decoder[Option[A]]): Decoder[A] =
+  OptionT(optionDecoder).getOrElse(defaultValue)
+```
+
+However, we can't do the same with our `CovariantOptionT`, because we can't even create an `OptionT` where the `F` type isn't covariant:
+
+```tut:book:fail
+def wrap[A](optionDecoder: Decoder[Option[A]]): CovariantOptionT[Decoder, A] =
+  CovariantOptionT[Decoder, A](optionDecoder)
+```
+
+In this particular case, `Decoder` _could_ be declared as covariant, but it's not. It would be unfortunate to lose the ability to use a monad transformer because a 3rd party library chose not to make a type covariant. And perhaps more importantly, sometimes you might want to use an `OptionT` with an `F` type that fundamentally isn't covariant in nature, such as `Monoid` (which is invariant) or `Order` (which is contravariant in nature and is declared as invariant in its type definition).
```

Comment:
Is this really a good argument without further context? At this point I think it makes sense to point out that `OptionT` also has use cases for contravariant type constructors, like a `contramap` function.

Reply:
Yeah, I had a whole section written up here that was using `Show` and then ran into https://github.com/typelevel/cats/issues/2545 😿 . I became hesitant to emphasize contravariant or invariant functors too much because in reality `OptionT` doesn't really support those well, and people who tried would run into issues. Also even assuming that those things existed, I was struggling to come up with examples using `OptionT` that didn't feel really contrived but didn't bring in too much other context. If you have have any good ideas, please give them. Also I may be able to just bring up the example that I have further down with `Eq`. I can look into rearranging a bit and providing better motivation.

Reply:
Fair point. In that case, I think a potentially weak argument is better than something that requires instances that don't exist yet :smile: